fast_str/strcasecmp.c

```diff
@@ -456,13 +456,13 @@ stricmp_avx2(const unsigned char *s1, const unsigned char *s2, size_t len)
 
 	while (i + 8 <= len) {
 		c |= lct[s1[i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
-		c |= lct[s1[++i]] ^ lct[s2[i]];
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
+		c |= (++i, lct[s1[i]] ^ lct[s2[i]]);
```

Comment:
@michael-grunder  thank for the PR and I'm glad that the blog post was useful for you!

May I know why the adjustment for the prefix increment are required? The new prefix increment seems doing just the same as the original one.

Reply:
I believe that `c |= lct[s1[++i]] ^ lct[s2[i]]` results in undefined behavior as the compiler is free to reorder the operations because `xor` is not a sequence point.

The code worked as-is for me on gcc and clang so perhaps this is not _actually_ a problem that happens in practice.  It's also possible that it is not undefined behavior in C++11.  I'm a C programmer (C++ confuses me :smile:).

Tiny example of what I'm seeing:

```bash
➜  scratch cat unsequenced.cc
#include <stdio.h>
int main() {
    int i = 0, v = 0;
    int arr1[] = {0, 1, 2};
    int arr2[] = {1, 2, 4};
    v |= arr2[arr1[++i]] ^ arr2[arr1[i]];
}
➜  scratch g++ -std=c++11 unsequenced.cc
unsequenced.cc:6:20: warning: unsequenced modification and access to 'i' [-Wunsequenced]
    v |= arr2[arr1[++i]] ^ arr2[arr1[i]];
```

Reply:
Oh, I see. I tried to compile exactly your example under Linux with G++ 6.3 and 7.2 and it compiles just fine, however Clang (g++ under MacOS) is more strict and shows the error. Thanks!